### PR TITLE
Design: namespace selection by label in resource policy

### LINKED ISSUE #9771: Design: ResourcePolicy-based namespace selection by label
Track design for ResourcePolicy-based namespace selection by label #7492

### PR DESCRIPTION
**Pull request by kaovilai**
## Summary

- Adds design document for extending `includeExcludePolicy` in ResourcePolicy ConfigMap with `includedNamespacesByLabel` and `excludedNamespacesByLabel` fields
- Users can dynamically include/exclude namespaces by Kubernetes label selectors without modifying `BackupSpec` or schedule specs
- Covers precedence rules, OR/AND selector semantics, observability via backup status, and rejected alternatives
- Cross-checked against the fine-grained backup/restore filter policies merged from issue #9448 (`clusterScopedFilterPolicy`, `namespacedFilterPolicies` in `internal/resourcepolicies/resource_policies.go`): confirms no field/key collisions, documents that a ConfigMap using these fields can't be reused for Restore or the global backup volume policies ConfigMap, and justifies the selector-string format vs. the merged `ResourceFilter`'s `map[string]string` convention

## Test plan

- [ ] Design review by maintainers
- [ ] Implementation to follow after design approval

Fixes #9771

> [!Note]
> Responses generated with Claude

### CODE REVIEW

**Comment by Lyndon-Li** [comment]
@kaovilai 
FYI of this issue https://github.com/velero-io/velero/issues/9448, @adam-jian-zhang is trying to making a new type of resource selection solution that supports nested and fine grained selecting of resources (for both namespaced and cluster scope resources), this solution will also be implemented in ResourcePolicy because: 1. the interfaces are complex and the existing filters cannot handle it; 2. as we discussed previously, it is the right direction to have the sophisticated policies into ResourcePolicy.

I think there are some overlaps for these two designs, let's discuss how we are going to handle them. cc @adam-jian-zhang  @reasonerjt

**Comment by kaovilai** [comment]
Oh hi adam.

**Comment by adam-jian-zhang** [comment]
hi tiger,
Yes, I am also working on the backup filters improvement, here is my design:
https://github.com/adam-jian-zhang/velero/blob/backup-filter-design/design/backup-filter-enhancement/fine-grained-backup-filters-design.md
I have not raised  PR yet, we are doing internal review first to make the design solid before proposing PR, we have been doing this for two rounds and are quite close to proposing PR.
I think the design can also cover the use cases presented in this PR, and many others. 
Let's discuss how to move forward with this.

**Comment by kaovilai** [comment]
Thank you Adam. From community meeting, it is my understanding that your design will not cover "namespace selection by label" requirement documented in https://github.com/velero-io/velero/issues/7492.

The requirement states that user can choose namespaces to backup similar to `velero backup create --included-namespaces` by using labelSelector on namespaces instead of specifying namespace names into backup spec.

It was suggested to use resource policy for that, hence I retitled the issue in 7492. I will keep my design 9772 here open/draft and try to make it compatible with your design.

**Comment by adam-jian-zhang** [comment]
@kaovilai , 
I look it more carefully, I think I only need a small adjustment to incorporate the namespace with labels use cases in my design. Basic ideas is adding  labels selector in the namespacedFilterPolicies to express the namespace selection requirement, something like:
```yaml
version: v1
namespacedFilterPolicies:
  - namespaceLabelSelector: # New to address the use cases for this design
      environment: production
    excludedNamespaceLabelSelector:  # New to address the use cases for this design
      skip-selective-backup: "true"
    resourceFilters:
      - kinds: [ConfigMap, Secret, Deployment]
        labelSelector:
          app: my-app
```

What do you think?
 I will update my design later to include this.

**Comment by kaovilai** [comment]
That works! Thank you

**Comment by adam-jian-zhang** [comment]
@kaovilai,
I have tried to incorporate the namespace label selector use cases into fine grained resource filter design, here is the draft design,
https://github.com/adam-jian-zhang/velero/blob/backup-filter-design-alternative/design/backup-filter-enhancement/fine-grained-backup-filters-design.md#use-labels-to-assign-namespaces-to-different-backup-schedules

But during the review we all thought it was too complicated. 
And I had to bend several validation rules and inverse the empty resourceFilters meaning to accommodate that.
I think it's better to keep those two designs separate, since it addresses different requirements.
Namespace label filters address global namespace selection, and fine-grained resource filters address per namespace fine tuning after the selection. It is totally possible to deliver them separately.
 cc @Lyndon-Li @reasonerjt

File: design/namespace-label-selector-in-resource-policy_design.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
+# Namespace Selection by Label in Resource Policy
+
+## Glossary & Abbreviation
+
+**Backup Filter**: The mechanism in Velero that determines which Kubernetes resources are collected from the cluster and written into the backup archive. Backup filters currently operate on four dimensions: namespace, resource type, label, and cluster scope.
```
Comment (adam-jian-zhang) [inline design/namespace-label-selector-in-resource-policy_design.md]: Add two spaces after each item, to make sure it renders properly into each line.

File: design/namespace-label-selector-in-resource-policy_design.md [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+5. No new controller-level incompatibility check is needed. `prepareBackupRequest` in `pkg/controller/backup_controller.go` already rejects any non-nil `IncludeExcludePolicy` combined with old-style filters (`collections.UseOldResourceFilters(request.Spec)`) regardless of which `IncludeExcludePolicy` fields are populated. Since the new fields live on the existing `IncludeExcludePolicy` struct, this check covers them automatically.
+6. Log the resolved namespace set at info level in `prepareBackupRequest`.
+7. Add unit tests for selector parsing (including the empty-string rejection), independent resolution of `resolvedIncludedByLabel`/`resolvedExcludedByLabel` (including a configured include selector that matches zero namespaces, and an exclude-only policy), and precedence rules (`labelIncludeActive` baseline branch and hard-exclusion ordering).
+8. Add E2E test: schedule with no `includedNamespaces`, label selector in resource policy, verify only labeled namespaces are backed up.
+
+## Open Issues
+
+- **AND vs OR between list entries**: This design uses OR (union). Should we support AND by allowing nested lists? Deferring to a future enhancement.
```
Comment (adam-jian-zhang) [inline design/namespace-label-selector-in-resource-policy_design.md]: I think we should not defer this to future enhancement, changing behavior later causes user surprise.

What about adding a new field to specify the relationship among label selectors, something like: 
`selectorLogic: "OR" | "AND"` (defaulting to `"OR"`).
 If we want to support AND selectors, it is something like:
```yaml
version: v1
includeExcludePolicy:
  labelSelectorLogic: "AND" # NEW: "OR" (default) or "AND"
  includedNamespacesByLabel:
    - "team=platform"
    - "env=prod"
```

File: design/namespace-label-selector-in-resource-policy_design.md [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+### Observability
+
+`backup.status.includedNamespaces` does not exist on `BackupStatus` today. Adding it is out of scope for the initial implementation — it would require a new `BackupStatus` field, CRD schema/deepcopy regeneration, and its own compatibility review, none of which this design specifies. Instead, the initial implementation logs the resolved namespace set (info level) during `prepareBackupRequest` so operators can see which namespaces were actually selected via `kubectl logs`/backup logs. Populating a status field is deferred as a follow-on (see [Open Issues](#open-issues)); if picked up later, it needs its own design pass covering the schema addition and back-compat impact.
```
Comment (adam-jian-zhang) [inline design/namespace-label-selector-in-resource-policy_design.md]: If a cluster has thousands of namespaces and the selector matches 2,000 of them, logging the full array at `Info` level will severely spam the Velero backup logs. 

Recommendation: Consider logging the *count* of resolved namespaces at the `Info` level, and log the actual list of names at the `Debug` level (or truncate the list in the Info log, e.g., `[ns-1, ns-2, ... and 1998 more]`).

File: design/namespace-label-selector-in-resource-policy_design.md (hunk 1 above)
Comment (kaovilai) [reply design/namespace-label-selector-in-resource-policy_design.md]: Fixed — added trailing double-spaces to each Glossary entry so they render as separate lines.

> [!Note]
> Responses generated with Claude

File: design/namespace-label-selector-in-resource-policy_design.md (hunk 2 above)
Comment (kaovilai) [reply design/namespace-label-selector-in-resource-policy_design.md]: Good point — implemented `labelSelectorLogic: "OR" | "AND"` on `includeExcludePolicy` (default `"OR"`) instead of deferring, matching your suggested shape. AND combines entries within `includedNamespacesByLabel` (and independently within `excludedNamespacesByLabel`) via intersection instead of union. Added a worked example and threaded it through `resolveNamespacesByLabel` and validation.

> [!Note]
> Responses generated with Claude

File: design/namespace-label-selector-in-resource-policy_design.md (hunk 3 above)
Comment (kaovilai) [reply design/namespace-label-selector-in-resource-policy_design.md]: Agreed, fixed — now logs the resolved-namespace *count* at info level and the full name list at debug level, so large clusters where a selector matches many namespaces won't spam backup logs by default.

> [!Note]
> Responses generated with Claude

**Review by Ralthos — COMMENTED** [review]
Reading this alongside #10210, which is deciding namespace precedence right now, and there is an interaction worth settling before either lands.

The design states:

> `BackupSpec.LabelSelector` and `BackupSpec.OrLabelSelectors` are unaffected — they continue to select individual resources, not namespaces.

That is not quite true on `main` today. `nsTracker.init` tracks a namespace when the `Namespace` object's own labels match `BackupSpec.LabelSelector`:

```go
if nt.singleLabelSelector != nil &&
    nt.singleLabelSelector.Matches(labels.Set(namespace.GetLabels())) {
    nt.track(namespace.GetName())
    continue
}
```

`OrLabelSelectors` has the same branch. So a backup's label selector already selects namespaces, not only resources within them, and there is a test asserting it does so even against an explicit exclude:

```go
name: "ns not included by IE filter, but included by labelSelector",
ie:   collections.NewNamespaceIncludesExcludes().Excludes("ns1"),
expectedTrackedNS: []string{"ns1"},
```

That came in with #7697, which describes the intent as a union: if the namespace filters and the selectors pick different namespaces, the tracker takes both.

The reason this matters here is that the merge order in Precedence and Interaction is subtractive:

```
effective = baseline
           − BackupSpec.ExcludedNamespaces
           − resolvedExcludedByLabel
```

Exclusion applies last and wins over everything, which is the opposite of the union rule #7697 implemented for `BackupSpec.LabelSelector`. So the design already assumes the precedence that #10210 is proposing to introduce, while today's code does the other thing.

Three ways that could resolve, and I do not have a view on which is right:

If #10210 lands, the two are consistent and the sentence about selectors being unaffected just needs a note that namespace-level selection now honours exclusion.

If #10210 is dropped as working-as-designed, this design gets a namespace admitted by `BackupSpec.LabelSelector` that it did not subtract, and the precedence section would need a case for it.

Or namespace-level selection could be treated as one concept regardless of where the selector is configured, with a single precedence rule covering `BackupSpec.LabelSelector`, `includedNamespacesByLabel` and the name filters together. That is more work but it leaves one rule to document instead of two that have to be kept in agreement.

For what it is worth, #10209 was mine and I have since argued on #10210 that the behaviour it reports may be intended, so I am not pushing for a particular outcome here. It is the fact that two precedence models are being decided in parallel that seems worth flagging.

**Comment by kaovilai** [comment]
@Ralthos the namespaces if any that comes without including #10210 IIUC is only the namespace object itself, not any other kube objects that reside in there.

**Comment by Ralthos** [comment]
Confirmed. `filterNamespaces` only ever drops resources whose `groupResource` is `Namespaces`,
so being tracked decides whether the `Namespace` object is written and nothing else. Contents
are collected separately, and `track()` in `getResourceItems` is already gated on
`NamespaceIncludesExcludes.ShouldInclude`.

One consequence for this design. Precedence and Interaction says `BackupSpec.LabelSelector` is
unaffected. True for resource selection, but the effective set is written into
`BackupSpec.IncludedNamespaces`, and `nsTracker.init` reads that same filter:

```go
if nt.namespaceFilter.IncludeEverything() &&
    (nt.singleLabelSelector != nil || len(nt.orLabelSelector) > 0) {
    continue
}
```

That guard is the #7105 fix. `includedNamespacesByLabel` with empty `IncludedNamespaces` is the
doc's primary use case, and it makes `IncludeEverything()` false, so the guard stops firing and
namespaces get tracked by name even when their labels do not match `BackupSpec.LabelSelector`.
Extra `Namespace` objects in the tarball, contents unchanged, which is what makes it easy to
miss.

Maybe that is the intended behaviour, but #7105 was filed about this exact surprise, so it seems
worth stating either way.

**Review by adam-jian-zhang — DISMISSED** [review]
LGTM

**Comment by kaovilai** [comment]
cc: @velero-io/maintainer

**Review by sseago — DISMISSED** [review]

**Comment by kaovilai** [comment]
Addressing @Ralthos comments

**Review by adam-jian-zhang — APPROVED** [review]
LGTM

**Review by sseago — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]
